### ISSUE TITLE
Half Precision Training may cause discrepancies of test results between training and test time for the same model 

### ISSUE THREAD

**Issue by zhixuany** [kenziyuliu/MS-G3D#54]
Thanks for the fantastic work! I am able to run the code and reproduce results.

However, when I try to train MS-G3D on my own data, I happen to find that in some cases the test results of a saved model after a certain epoch does not match with its test results during training after that epoch. In my case, "Mean test loss" is NaN during training but is normal during test (load saved model and test). I end up find that this difference comes from the use of half precision training, since "--half" is only used for training which seems to change the model behavior a bit even though the model weights should be the same under both settings. Do you have any thoughts or suggestions on this?

By the way, I find that the apex workflow used here is not consistent with that recommended in the official github of apex (e.g. it recommends "calling the load_state_dict methods after amp.initialize" instead). Is there a specific reason for that?

**Comment by kenziyuliu** [comment]
Hi @zhixuany,

Thanks for your interest!

> Do you have any thoughts or suggestions on this?

So `apex` was simply used to cut memory usage (the training didn't quite fit in the memory with 1080Ti's); if resources permit I would suggest just remove half precision training. I haven't run into the issue you're describing and unfortunately don't have too many suggestions.

> ... Is there a specific reason for that?

No specific reasons that I could remember :). I believe the `apex` has evolved quite a bit since I wrote the code (2019) and best practices may have changed since then. Please feel free to follow the best practices suggested by `apex`.

**Comment by zhixuany** [comment]
Got it, thanks!
